### PR TITLE
CamelCaseVariableName: allow-underscore implemented

### PR DESCRIPTION
**Pull request by silentx1981**
Sorry for the new Request, but i was forked from manuelpichler/phpmd.

### CODE REVIEW

**Comment by silentx1981** [comment]
Why the test ist failed? (Sorry for my ignorance)

**Comment by silentx1981** [comment]
2016-04-17_18h41_10
 I don't understand the error, because the file exists?

**Comment by ravage84** [comment]
The test medthod is named ``testRuleDoesApplyForVariableNameWithUnderscore`` but the test file and it's class name is called ``testRuleDoesApplyForVariableNameWithUnderscores`` with a trailing ``s``.
They need to match.

File: src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php [hunk 1]
```diff
@@ -83,14 +83,20 @@ class CamelCaseVariableName extends AbstractRule implements MethodAware, Functio
      */
     public function apply(AbstractNode $node)
     {
+        $allowUnderscore = $this->getBooleanProperty('allow-underscore');
+        $pattern = '/^\$[a-zA-Z][a-zA-Z0-9]*$/';
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php]: Shouldn't it be ``/^\$[a-z][a-zA-Z0-9]*$/``? I think the new one allows capital letters as first letter.

File: src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php [hunk 2]
```diff
@@ -83,14 +83,20 @@ class CamelCaseVariableName extends AbstractRule implements MethodAware, Functio
...
+        if ($allowUnderscore === true) {
+            $pattern = '/^\$[_][a-zA-Z][a-zA-Z0-9]*$/';
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php]: Shouldn't it be ``/^\$[_][a-z][a-zA-Z0-9]*$/``? I think the new one allows capital letters as first letter after the underscore.

File: src/main/resources/rulesets/controversial.xml [hunk 3]
```diff
@@ -144,7 +144,15 @@ It is considered best practice to use the camelCase notation to name variables.
             ]]>
         </description>
         <priority>1</priority>
-        <properties />
+        <properties>
+            <property name="allow-underscore"
+                      description="Allow an optional, single underscore at the beginning."
+                      value="false" />
+            <property name="allow-underscore-test"
```
Comment (ravage84) [inline src/main/resources/rulesets/controversial.xml]: Could you please move that to a separate PR? It is untested, anyway. Plus the other rules don't have this.

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForVariableNameWithCapital.php [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesNotApplyForVariableNameWithCapital
+{
+    public $AlsoValidVariableName;
+}
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForVariableNameWithCapital.php]: Please add white line at the end.

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed.php [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed
+{
+    public $_camelCaseVariableNameWithUnderscoreAtTheBeginning;
+}
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed.php]: Please add white line at the end.

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForVariableNameWithUnderscores.php [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?php
+class testRuleDoesApplyForVariableNameWithUnderscores
+{
+    public function variableNameWithUnderscores()
+    {
+        $not_camel_case_variable_name = 'foo';
+    }
+}
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForVariableNameWithUnderscores.php]: Please add white line at the end.

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed.php [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed
+{
+    public $_camelCaseVariableNameWithUnderscoreAtTheBeginning;
+}
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed.php]: Please add white line at the end.

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed.php [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesApplyForValidVariableNameWithUnderscoreWhenNotAllowed.php]: Is there a test case for this?

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithNoUnderscoreWhenAllowed.php [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesNotApplyForValidVariableNameWithNoUnderscoreWhenAllowed
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithNoUnderscoreWhenAllowed.php]: Is there a test case for this?

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed.php [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForValidVariableNameWithUnderscoreWhenAllowed.php]: Is there a test case for this?

File: src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForVariableNameWithCapital.php [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<?php
+class testRuleDoesNotApplyForVariableNameWithCapital
```
Comment (ravage84) [inline src/test/resources/files/Rule/Controversial/CamelCaseVariableName/testRuleDoesNotApplyForVariableNameWithCapital.php]: Is there a test case for this?

File: src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php [hunk 12]
```diff
@@ -83,14 +83,20 @@ class CamelCaseVariableName extends AbstractRule implements MethodAware, Functio
      */
     public function apply(AbstractNode $node)
```
Comment (eeree) [inline src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php]: I think, that this change should be applied to `CamelCasePropertyName` as even tests you provided covers only properties - not variables.

Comment (ravage84) [inline src/main/php/PHPMD/Rule/Controversial/CamelCaseVariableName.php]: ``CamelCasePropertyName`` already provides the ``allow-underscore`` property, see https://github.com/phpmd/phpmd/blob/2.6.0/src/main/php/PHPMD/Rule/Controversial/CamelCasePropertyName.php#L68

I think the @silentx1981 copied the tests from the prior PR (mostly) 1:1.

**Comment by ravage84** [comment]
@eeree if you want to take care of this, feel free 😼

**Comment by ravage84** [comment]
Replaced by https://github.com/phpmd/phpmd/pull/749
